File: tests/bindings/dotnet/foo.Tests/CallbackTest.cs
using System;
using Xunit;
using foo;
using System.Collections.Generic;

namespace foo.Tests
{
    class CallbackImpl : ICallbackInterface
    {
        public uint lastValue = 0;
        public TimeSpan lastDuration = TimeSpan.MinValue;
        public Names name = null;
        public System.Collections.Generic.IList<Names> names = new System.Collections.Generic.List<Names>();

        public uint OnValue(uint value)
        {
            lastValue = value;
            return value;
        }

        public TimeSpan OnDuration(TimeSpan value)
        {
            lastDuration = value;
            return value;
        }

        public void OnNames(Names names)
        {
            this.name = names;
        }

        void ICallbackInterface.OnSeveralNames(ICollection<Names> names)
        {
            foreach(var name in names)
            {
                this.names.Add(name);
            }
        }
    }

    class CallbackFinalizerCounterImpl : ICallbackInterface
    {
        private readonly Counters counters;

        public uint OnValue(uint value)
        {
            return value;
        }

        public TimeSpan OnDuration(TimeSpan value)
        {
            return value;
        }

        public CallbackFinalizerCounterImpl(Counters counters)
        {
            this.counters = counters;
            this.counters.numConstructorsCalled++;
        }

        public void OnNames(Names names)
        {
            
        }

        void ICallbackInterface.OnSeveralNames(ICollection<Names> names)
        {

        }

        ~CallbackFinalizerCounterImpl()
        {
            this.counters.numFinalizersCalled++;
        }        
    }

    class Counters
    {
        public int numConstructorsCalled = 0;
        public int numFinalizersCalled = 0;
    }

    public class CallbackTests
    {
        [Fact]
        public void CallbackTest()
        {            
            using var cbSource = new CallbackSource();
            
            var cb = new CallbackImpl();
            cbSource.SetInterface(cb);

            Assert.Equal(0u, cb.lastValue);
            var result = cbSource.SetValue(76);
            Assert.Equal(76u, result);
            Assert.Equal(76u, cb.lastValue);

            Assert.Equal(TimeSpan.MinValue, cb.lastDuration);
            var timeResult = cbSource.SetDuration(TimeSpan.FromSeconds(76));
            Assert.Equal(TimeSpan.FromSeconds(76), timeResult);
            Assert.Equal(TimeSpan.FromSeconds(76), cb.lastDuration);

            Assert.Null(cb.name);
            cbSource.InvokeOnNames(new Names("John", "Smith"));
            Assert.Equal("John", cb.name.FirstName);
            Assert.Equal("Smith", cb.name.LastName);

            Assert.Empty(cb.names);
            cbSource.InvokeOnSeveralNames();
            Assert.Equal(2, cb.names.Count);
            Assert.Equal("jane", cb.names[0].FirstName);
            Assert.Equal("doe", cb.names[0].LastName);
            Assert.Equal("jake", cb.names[1].FirstName);
            Assert.Equal("sully", cb.names[1].LastName);
        }

        private void SingleRun(Counters counters)
        {
            using var cbSource = new CallbackSource();
            cbSource.SetInterface(new CallbackFinalizerCounterImpl(counters));
            cbSource.SetValue(76);
        }

        [Fact]
        public void CallbackMemoryLeakTest()
        {
            var counters = new Counters();
            var numRuns = 1000;

            for (int i = 0; i < numRuns; i++)
            {
                SingleRun(counters);
            }

            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();

            Assert.Equal(numRuns, counters.numConstructorsCalled);
            Assert.Equal(numRuns, counters.numFinalizersCalled);
        }
    }
}

File: tests/bindings/java/foo-tests/src/test/java/io/stepfunc/foo_test/CallbackTest.java
package io.stepfunc.foo_test;

import io.stepfunc.foo.CallbackInterface;
import io.stepfunc.foo.CallbackSource;
import io.stepfunc.foo.Names;
import org.assertj.core.data.Percentage;
import org.joou.UInteger;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Disabled;

import java.time.Duration;
import java.util.ArrayList;

import static org.assertj.core.api.Assertions.assertThat;
import static org.joou.Unsigned.uint;

public class CallbackTest {
    static class CallbackImpl implements CallbackInterface {
        public UInteger lastValue = uint(0);
        public Duration lastDuration = null;

        public Names name = null;

        public java.util.List<Names> names = new ArrayList<>();


        @Override
        public UInteger onValue(UInteger value) {
            this.lastValue = value;
            return value;
        }

        @Override
        public Duration onDuration(Duration value) {
            this.lastDuration = value;
            return value;
        }

        @Override
        public void onNames(Names names) {
            this.name = names;
        }

        @Override
        public void onSeveralNames(java.util.List<Names> names) {
            this.names.addAll(names);
        }
    }

    @Test
    public void InterfaceAndOneTimeCallbackTest() {
        try(CallbackSource cbSource = new CallbackSource()) {
            CallbackImpl cb = new CallbackImpl();
            cbSource.setInterface(cb);

            assertThat(cb.lastValue).isEqualTo(uint(0));
            assertThat(cbSource.setValue(uint(76))).isEqualTo(uint(76));
            assertThat(cb.lastValue).isEqualTo(uint(76));

            assertThat(cb.lastDuration).isNull();
            assertThat(cbSource.setDuration(Duration.ofSeconds(76))).isEqualTo(Duration.ofSeconds(76));
            assertThat(cb.lastDuration).isEqualTo(Duration.ofSeconds(76));

            assertThat(cb.name).isNull();
            cbSource.invokeOnNames(new Names("John", "Smith"));
            assertThat(cb.name.firstName).isEqualTo("John");
            assertThat(cb.name.lastName).isEqualTo("Smith");

            assertThat(cb.names.isEmpty()).isTrue();
            cbSource.invokeOnSeveralNames();
            assertThat(cb.names.size()).isEqualTo(2);
            assertThat(cb.names.get(0).firstName).isEqualTo("jane");
            assertThat(cb.names.get(0).lastName).isEqualTo("doe");
            assertThat(cb.names.get(1).firstName).isEqualTo("jake");
            assertThat(cb.names.get(1).lastName).isEqualTo("sully");
        }
    }

    static class CallbackFinalizerCounterImpl implements CallbackInterface {
        private final Counters counters;

        @Override
        public UInteger onValue(UInteger value) {
            return value;
        }
        @Override
        public Duration onDuration(Duration value) {
            return value;
        }

        @Override
        public void onNames(Names names) {}

        @Override
        public void onSeveralNames(java.util.List<Names> names) {}

        public CallbackFinalizerCounterImpl(Counters counters) {
            this.counters = counters;
            this.counters.numConstructorsCalled++;
        }

        @Override
        protected void finalize() {
            this.counters.numFinalizersCalled++;
        }
    }

    static class Counters {
        public int numConstructorsCalled = 0;
        public int numFinalizersCalled = 0;
    }
}
